File: test_errors.py
#!/usr/bin/env python3
"""
Test mpc_designation.py error handling.

Tests various classes of invalid input to ensure proper error detection.
Reads test cases from error_test_cases.csv.

Usage: python3 test_errors.py [error_test_cases.csv]
"""

import sys
import re
import codecs
from mpc_designation import convert_simple, MPCDesignationError


def unescape_string(s: str) -> str:
    """Parse escape sequences in a string."""
    # Handle \xNN hex escapes first
    def hex_replace(match):
        return chr(int(match.group(1), 16))

    s = re.sub(r'\\x([0-9a-fA-F]{2})', hex_replace, s)

    # Handle standard escapes
    s = s.replace('\\n', '\n')
    s = s.replace('\\r', '\r')
    s = s.replace('\\t', '\t')
    s = s.replace('\\f', '\f')
    s = s.replace('\\v', '\v')
    s = s.replace('\\0', '\0')
    s = s.replace('\\\\', '\\')

    return s


def run_error_tests(csv_file: str) -> bool:
    """Run error tests from CSV file."""
    total = 0
    passed = 0
    failed = 0

    print("=== MPC Designation Error Tests ===\n")

    with open(csv_file, 'r', encoding='utf-8') as fp:
        for line_num, line in enumerate(fp, 1):
            line = line.strip()

            # Skip empty lines and comments
            if not line or line.startswith('#'):
                continue

            # Skip header
            if line.startswith('category,'):
                continue

            # Parse CSV (simple split, no quotes handling needed for this data)
            parts = line.split(',', 4)
            if len(parts) < 5:
                continue

            category = parts[0]
            subcategory = parts[1]
            input_str = parts[2]
            expected_error = parts[3]
            description = parts[4]

            # Unescape the input string
            input_str = unescape_string(input_str)

            total += 1

            # Run the test
            try:
                output = convert_simple(input_str)
                got_error = False
            except MPCDesignationError as e:
                got_error = True
                error_msg = str(e)
            except Exception as e:
                got_error = True
                error_msg = str(e)

            test_passed = False

            if expected_error == "valid":
                # Expect success
                if not got_error:
                    test_passed = True
                else:
                    print(f"FAIL [{category}/{subcategory}]: '{description}'")
                    print(f"      Expected: valid conversion")
                    print(f"      Got:      {error_msg}")
                    failed += 1
            else:
                # Expect error
                if got_error:
                    test_passed = True
                else:
                    print(f"FAIL [{category}/{subcategory}]: '{description}'")
                    print(f"      Expected: error ({expected_error})")
                    print(f"      Got:      '{output}' (success)")
                    failed += 1

            if test_passed:
                passed += 1

    print("\n=== Error Test Results ===")
    print(f"Total:  {total}")
    print(f"Passed: {passed}")
    print(f"Failed: {failed}")

    return failed == 0


def main():
    csv_file = sys.argv[1] if len(sys.argv) > 1 else "error_test_cases.csv"

    try:
        success = run_error_tests(csv_file)
        sys.exit(0 if success else 1)
    except FileNotFoundError:
        print(f"Error: Cannot open file: {csv_file}", file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: rust/src/bin/test_errors.rs
//! Error Test Runner
//!
//! Test mpc_designation error handling against CSV file of invalid inputs.
//!
//! Usage: test_errors <csv_file>

use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::process;
use mpc_designation::convert_simple;

/// Parse escape sequences in a string.
fn unescape_string(s: &str) -> String {
    let mut result = String::new();
    let mut chars = s.chars().peekable();

    while let Some(c) = chars.next() {
        if c == '\\' {
            match chars.next() {
                Some('x') => {
                    // Handle \xNN hex escape
                    let hex: String = chars.by_ref().take(2).collect();
                    if hex.len() == 2 {
                        if let Ok(byte) = u8::from_str_radix(&hex, 16) {
                            result.push(byte as char);
                            continue;
                        }
                    }
                    result.push_str("\\x");
                    result.push_str(&hex);
                }
                Some('n') => result.push('\n'),
                Some('r') => result.push('\r'),
                Some('t') => result.push('\t'),
                Some('0') => result.push('\0'),
                Some('\\') => result.push('\\'),
                Some(other) => {
                    result.push('\\');
                    result.push(other);
                }
                None => result.push('\\'),
            }
        } else {
            result.push(c);
        }
    }

    result
}

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 {
        eprintln!("Usage: test_errors <csv_file>");
        process::exit(1);
    }

    let csv_file = &args[1];

    let file = File::open(csv_file).unwrap_or_else(|e| {
        eprintln!("Cannot open file {}: {}", csv_file, e);
        process::exit(1);
    });

    let reader = BufReader::new(file);

    let mut total = 0u64;
    let mut passed = 0u64;
    let mut failed = 0u64;

    println!("=== MPC Designation Error Tests ===\n");

    for line_result in reader.lines() {
        let line = match line_result {
            Ok(l) => l,
            Err(_) => continue,
        };

        let line = line.trim();

        // Skip empty lines and comments
        if line.is_empty() || line.starts_with('#') {
            continue;
        }

        // Skip header
        if line.starts_with("category,") {
            continue;
        }

        // Parse CSV (simple split)
        let parts: Vec<&str> = line.splitn(5, ',').collect();
        if parts.len() < 5 {
            continue;
        }

        let category = parts[0];
        let subcategory = parts[1];
        let input_str = unescape_string(parts[2]);
        let expected_error = parts[3];
        let description = parts[4];

        total += 1;

        // Run the test
        let result = convert_simple(&input_str);
        let got_error = result.is_err();

        let test_passed = if expected_error == "valid" {
            // Expect success
            if got_error {
                println!("FAIL [{}/{}]: '{}'", category, subcategory, description);
                println!("      Expected: valid conversion");
                println!("      Got:      {}", result.unwrap_err());
                failed += 1;
                false
            } else {
                true
            }
        } else {
            // Expect error
            if got_error {
                true
            } else {
                println!("FAIL [{}/{}]: '{}'", category, subcategory, description);
                println!("      Expected: error ({})", expected_error);
                println!("      Got:      '{}' (success)", result.unwrap());
                failed += 1;
                false
            }
        };

        if test_passed {
            passed += 1;
        }
    }

    println!("\n=== Error Test Results ===");
    println!("Total:  {}", total);
    println!("Passed: {}", passed);
    println!("Failed: {}", failed);

    process::exit(if failed > 0 { 1 } else { 0 });
}
